### DISCUSSION TITLE
Which CI environment integration should we cover next?

### DISCUSSION THREAD

**Discussion by gerhard** [dagger/dagger#1677] [General]
We started writing a new doc page - **From local dev to CI environment** https://github.com/dagger/dagger/pull/1678 - and would like to know which CI/CD platform you would find most useful for us to document next.

Since everyone that is using Dagger today already has a GitHub account (one of the early access requirements), starting with GitHub Actions seemed like a most reasonable first step.

Drop a comment below with the CI/CD platform that you use. We intend to document the one with the most mentions next.

**Comment by gerhard** [comment]
CircleCI: https://github.com/thechangelog/changelog.com/blob/a8f472230fe8cfc2e599523b4a6340f384607d16/.circleci/config.yml

**Reply by maert** [reply]
Dagger orb (CircleCI's lego brick) here: https://github.com/logicalhq/dagger-orb
Not yet listed as a public orb in the Orb Registry: https://circleci.com/developer/orbs

**Reply by jpadams** [reply]
@gcoguiec It looks like your `dagger-orb` repo isn't up anymore. Would you be willing to donate the code to Dagger to maintain?

**Reply by jpadams** [reply]
Have a CircleCI example going up in PR right now. With more complex example (caching) to follow. https://github.com/dagger/dagger/issues/2842

**Comment by helderco** [comment]
I use GitLab. I have private repos on GitLab, with dagger configs. Didn't have time to make dagger work with it, but I'm able to test it.

**Reply by gerhard** [reply]
Thank you @adrienbrault for contributing this in https://github.com/dagger/dagger/pull/1949 🙌🏻

**Reply by nicolas-goudry** [reply]
What about GitLab CI with self-hosted runners **without** privilege escalation capabilities?

My point being, is there a way to run Dagger in GitLab CI without exposing the runner Docker daemon?

**Reply by adrienbrault** [reply]
@nicolas-goudry you could try sysbox: https://blog.nestybox.com/2020/10/21/gitlab-dind.html

**Reply by nicolas-goudry** [reply]
@adrienbrault Thanks, I’ll take a look at this and maybe give it a try with Dagger :slightly_smiling_face:

**Reply by loewenstein** [reply]
Is that a "no" for using dagger without exposing the Docker daemon?
I would suppose that most Kubernetes based CI systems will recommend not exposing the Docker daemon.

**Reply by maudeFW** [reply]
@nicolas-goudry Did you try it already? If yes, what have been your experiences?
I'm evaluating different CI/CD Frameworks to run on GitLab CI Kubernetes Executor so input would be helpful, thanks.

**Reply by nicolas-goudry** [reply]
@maudeFW Sorry for replying so late, I didn’t have time to look into it yet…

**Reply by jpadams** [reply]
GitLab example getting better and better here: https://docs.dagger.io/1201/ci-environment

**Reply by Aboitier** [reply]
Any updated doc about integrating/migrating from gitlab-CI to dagger ?

**Reply by marcosnils** [reply]
hey @Aboitier! Dagger won't replace your CI runner, but we have several people in our Discord using both gitlab hosted and self-hosted runners with success. If you have some questions about it, would love to chat about it more in our Discord server :pray:

**Comment by TomChv** [comment]
At Epitech, there was courses on Jenkins, I'm not sure it's relevant but some infrastructures are still using it so maybe we can do something with Dagger

**Reply by reitzig** [reply]
Jenkins is still prevalent in my customer projects, so that would definitely be an important target.

**Reply by thatsk** [reply]
Me too

**Reply by solarsnake** [reply]
I'll third for jenkins

**Reply by kenyee** [reply]
4th for jenkins

**Reply by itzikiusa** [reply]
Me to

**Reply by bobhenkel** [reply]
Me too

**Reply by jpadams** [reply]
Jenkins example and video here: https://docs.dagger.io/1201/ci-environment

**Comment by maert** [comment]
GitHub knows which are the big CI tools on their platform (or competitive) and has transition guides from each one to GitHub Actions. So CircleCI, TravisCI, Jenkins, and GitLab CI/CD all make sense. Azure Pipelines (from Azure DevOps) could also make sense. Long term vision of MSFT seems to be GitHub & Actions, but a long tail of ADO users on prem and in Azure.

https://docs.github.com/en/actions/migrating-to-github-actions

Also makes sense to support the cloud's tools:
- AWS - [Code Pipeline](https://aws.amazon.com/codepipeline/)/[Code Build](https://aws.amazon.com/codebuild/)
- Azure - [Azure Pipelines](https://azure.microsoft.com/en-us/services/devops/pipelines/)
- GCP - [Cloud Build](https://cloud.google.com/build)
- Alibaba Cloud - [Cloud DevOps Pipeline (Flow)](https://www.alibabacloud.com/product/apsara-deveops/flow)
- Tencent Cloud - [CODING](https://intl.cloud.tencent.com/products/coding-cd)
- Digital Ocean - doesn't seem to have own tool. Lots of articles from folks using above tools to build for and deploy to DO.

**Reply by adriano-campanhola** [reply]
Taking into account your argument regarding Azure Pipelines, dagger should find a good use case for teams that want to slowly transition from ADO to Github.

**Reply by derekrprice** [reply]
Yes, we're on Azure DevOps Pipelines.  Would like to see this.

**Reply by StephenStrickland** [reply]
My team is also on ADO, would love to see this as well

**Reply by fafg** [reply]
Companies like Daimler (Mercedes-Benz), BMW, Volkswagen) and many others big would like to see this integrated with Azure DevOps Pipelines.

**Reply by Clean-Cole** [reply]
AZDO support is required for better or for worse.

**Reply by samclusker** [reply]
There's a lot of cloud native pipeline adoption. Personally I'd like to see support for ADO build pipelines.

**Reply by joachim-isaksson** [reply]
I have a semi-complete simple implementation for azure devops at https://github.com/joachim-isaksson/dagger-for-azure-devops 

It's based on the existing github actions implementation. Sadly I don't have the time to finish for the next few days, so if someone is so inclined, feel free to take it over :)

The task runs, dagger downloads and installs, runs correctly and the task fails in cleanup (a recursive delete that won't)

**Reply by brewsteropsdev** [reply]
I too have a large customer on ADO - would like to leverage there.

**Reply by invisibleaxm** [reply]
my vote for Azure DevOps as well.

**Reply by DavidNavalho** [reply]
Chiming in for ADO as well, as I've seen more and more clients using those.

**Reply by tadavid-cae** [reply]
Please make Azure Devops a reality!!!

**Reply by Pre-Z** [reply]
This would be a game changer for Azure DevOps Pipelines.

**Reply by tadavid-cae** [reply]
@gerhard  can you tell us the probability that you will integrate dagger with azure devops? :)

**Reply by jpadams** [reply]
@tadavid-cae @Pre-Z What would ADO support look like?

Would there need to be an extension or just an Azure Pipelines YAML example?

**Comment by gerhard** [comment]
Tekton: https://github.com/dagger/dagger/issues/1655

**Reply by maert** [reply]
And by extension, [Jenkins X](https://jenkins-x.io) that uses Tekton inside.

**Reply by jpadams** [reply]
We have Tekton here :) https://docs.dagger.io/1201/ci-environment

**Comment by maert** [comment]
[TeamCity](https://www.jetbrains.com/teamcity/) by JetBrains. Could create a [Kotlin script wrapper](https://www.youtube.com/watch?v=mnflFBtc1WY) for Dagger.

**Reply by darkn3rd** [reply]
This one is important for my organization.  Also, TeamCity supports inheritance, while Jenkins and others do not.  

Example: root project (with secrets to access git and triggers for pull request) has a child of build artifacts, which builds binaries on various platforms, and this has test children pipelines for integration tests and unit tests that consume the artifacts, so `root > build > {unit_tests, integration_tests}`.



, so for example, integration test pipeline and unit test pipeline could be dependent on a parent build pipeline that is executed prior and stores executable artifacts, and this in term is inherited from a parent that fetches the code and has secrets used to access the gitrepo, as well as triggers for pull requests, and so on.

**Reply by jpadams** [reply]
@darkn3rd if you have something that could help start a simple Team City example (or links), that would be great. Also PRs accepted ;)

Inspiration: https://docs.dagger.io/1201/ci-environment

**Reply by pavel-krizskii** [reply]
Info about TeamCity integration has been added to the docs: https://docs.dagger.io/getting-started/ci-integrations/teamcity

**Comment by eddumelendez** [comment]
[ArgoCD](https://argoproj.github.io/cd/)

**Reply by marcosnils** [reply]
Argo is mostly a CD platform, isn't it? It already relies on pre-existing CI solutions like GA, Circle, Tekton, etc for the CI part. Not sure what could be done in Dagger as an Argo direct integration :thinking:

**Reply by demasir** [reply]
@marcosnils I'm not sure if that's what @eddumelendez meant, but we use Argo Workflows to create our CI pipelines. It could be useful to have Dagger integrated with it so we could test them locally. Argo Workflows + Argo Events + Argo CD is basically a CI/CD platform by itself. For reference: https://argoproj.github.io/argo-workflows/

**Reply by marcosnils** [reply]
got it, thx! Wasn't aware about workflows. It looks like a potential opportunity for dagger as well.

**Comment by hihellobolke** [comment]
Jenkins - most enterprises still use it.. multi man-decades of tooling are around it.

**Reply by jpadams** [reply]
Jenkins example and video here: https://docs.dagger.io/1201/ci-environment

**Comment by eddumelendez** [comment]
[Concoruse CI](https://concourse-ci.org/)

**Reply by CorentinDy** [reply]
[Concourse CI](https://concourse-ci.org/)

**Reply by radiosilence** [reply]
Concourse would be amazing, seeing as it's modern, built for CD, able to handle extreme complexity and is not tied to a provider.

**Comment by manjunathb4461** [comment]
Are there any plans to implement the same for bitbucket pipelines?

**Reply by padraic-padraic** [reply]
This would interest me a lot, and help when it comes to pushing dagger at work!

**Reply by Lp-Francois** [reply]
Same here!

**Reply by PDickten** [reply]
Same here!

**Reply by dancordi** [reply]
same to me

**Reply by manjunathb4461** [reply]
@gerhard I created a simple [dagger pipe](https://github.com/manjunathb4461/dagger-pipe) for bitbucket pipelines just like [dagger-for-github](https://github.com/dagger/dagger-for-github)
But ended up with following error
```
{"level":"error","error":"exit status 125","output":"docker: Error response from daemon: authorization denied by plugin pipelines: --privileged=true is not allowed.\nSee 'docker run --help'.\n","time":"2022-04-03T07:20:49Z","time":"2022-04-03T07:20:49Z","caller":"/home/runner/work/dagger/dagger/util/buildkitd/buildkitd.go:210","message":"unable to start buildkitd"}
```
since bitbucket pipelines doesn't allow container(pipe in this case) to be run in `--privileged` mode, which is a requirement for [buildkit](https://github.com/dagger/dagger/blob/main/util/buildkitd/buildkitd.go#L196). Github actions also doesn't allow privileged mode by default, so wondering how it was solved for github actions. I would be happy to receive your feedback regarding the same.

**Reply by joachim-isaksson** [reply]
Ended up with the same error that I couldn't get around. I may be a bit off, but I think it may be possible to get it running once this is fixed.

https://jira.atlassian.com/browse/BCLOUD-17590

**Reply by dnalbach** [reply]
Bitbucket Pipeline docs say they do not allow `privileged` mode.
https://support.atlassian.com/bitbucket-cloud/docs/run-docker-commands-in-bitbucket-pipelines/

A potential workaround would be to use Bitbucket Runners and self-host the build machine, at which point you have total control over it, though you do lose the ephemeral aspect. I'm going to try that for some of our projects.

**Reply by jpadams** [reply]
Seems to still not be possible with Atlassian-hosted runners at the moment, even though BuildKit support is available in preview, since `--privileged=true` is still not supported
https://jira.atlassian.com/browse/BCLOUD-13420
and we currently run BuildKit as `--privileged` in Dagger.
https://github.com/dagger/dagger/blob/main/util/buildkitd/buildkitd.go#L238

<img width="1726" alt="Screen Shot 2022-07-20 at 11 34 35 PM" src="https://user-images.githubusercontent.com/3187222/180145974-31898289-810e-4006-bcc5-478f2c8517a2.png">

Seems like self-hosted runners with dind or a DOCKER_HOST over ssh would be needed.

```
image: atlassian/default-image:3

pipelines:
  default:
     - step:
         name: 'Install and run Dagger'
         script:
           - cd /usr/local
           - wget -O - https://dl.dagger.io/dagger/install.sh | sh
           - cd -
           - dagger version
           - dagger project update
           - DOCKER_BUILDKIT=1 dagger do hello --log-format plain
         services:
          - docker
```

**Comment by marcosnils** [comment]
My personal experience tells me to take a look at Jenkins / Jenkins X (and Tekton in consequence). Most of the times I've seen Jenkins pipeline using docker plugins, they were super difficult to follow and configure. I have a feeling that there's a big opportunity there to showcase a simpler and more efficient solution that integrates with those tools like Dagger.

**Reply by jpadams** [reply]
Jenkins and Tekton here :)
https://docs.dagger.io/1201/ci-environment

**Comment by ejunker** [comment]
[AWS CodeBuild](https://aws.amazon.com/codebuild/)

**Reply by Jojo134** [reply]
A basic AWS Codebuild was possible using a build task configured with the `aws/codebuild/standard:5.0` image and the following Buildspec, `pwd, ls` are still in there to see the folders/structure in the logs:
```
version: 0.2

phases:
  pre_build:
    commands:
      - pwd
      - curl -L https://dl.dagger.io/dagger/install.sh | sh
      - ls
      - ./bin/dagger version
  build:
    commands:
       - ./bin/dagger project init
       - ./bin/dagger project update 
       - ./bin/dagger do build
```

**Comment by KKcorps** [comment]
Jenkins

**Comment by cuken** [comment]
ADOC

**Comment by dieend** [comment]
Buildkite

**Comment by bmeverett** [comment]
GCP - [Cloud Build](https://cloud.google.com/build)

**Comment by shainegordon** [comment]
Bitbucket pipelines

**Comment by jenfredwell** [comment]
Jenkins!

**Comment by gaui** [comment]
Azure Pipelines

**Comment by technotenshi** [comment]
CircleCi please

**Comment by nrfrush** [comment]
Argo Workflows

**Comment by adambonny** [comment]
Gitlab

**Comment by darrenparkinson** [comment]
Another vote for gitlab please.

**Comment by adrienbrault** [comment]
Hey, I've just opened a PR with a sample gitlab ci configuration: https://github.com/dagger/dagger/pull/1949

**Comment by dido18** [comment]
Jenkins

**Comment by edromero** [comment]
Codefresh

**Reply by edromero** [reply]
Is there an official dagger docker image? I can work on creating the Codefresh pipeline with the existing todoapp example

**Comment by goncalobessa** [comment]
Jenkins dagger#1677

**Comment by ghishadow** [comment]
Sourcehut CI https://sourcehut.org/

**Comment by perfectmak** [comment]
CircleCI

**Comment by WiggleThb** [comment]
Jenkins

**Comment by Ara4Sh** [comment]
Jenkins

**Comment by MikeMichel** [comment]
Jenkins

**Comment by zzzkl** [comment]
Jenkins

**Comment by velkjaer** [comment]
🗡️  Tekton 🗡️

**Comment by ozbillwang** [comment]
drone.io

**Comment by ozbillwang** [comment]
After compare the exist solutions for Github Action and Gitlab CI (currently others are not ready) in https://docs.dagger.io/1201/ci-environment, I just realized, the concept that dagger can run on all cicd, can be easily replaced by a `Makefile`

when change from `dagger do build` to `make do-build`, it works on all pipelines, no need write different configuration for different cicd tools at all.

```
$ cat Makefile

...
do-build:
        echo "make sure docker run"
        echo "make sure dagger installed, or run a docker image as agent, which has dagger installed."
        dagger do build
```

LoL

**Comment by adyanth** [comment]
I'm surprised at how [DroneCI](https://www.drone.io) still hasn't made the list.

**Reply by ozbillwang** [reply]
it was listed, 6 hours before yours by me :-)

**Reply by OneCricketeer** [reply]
It was https://github.com/dagger/dagger/discussions/1677#discussioncomment-2477166

**Reply by adyanth** [reply]
Oh LOL, my bad. I searched with DroneCI, I see you have added drone.io :)

**Comment by cemozer** [comment]
[GoCD](https://www.gocd.org/) . It's kind of a CD tool but relies on docker agents for builds. So I assume shell based integration will work.

**Reply by adyanth** [reply]
I'm so sad something called GoCD does not run on Go, and more so that it runs on Java :)

**Comment by OneCricketeer** [comment]
IMO, this discussion should be locked. Lots of duplicate comments, and mostly all the ones listed here have been mentioned. 

https://landscape.cd.foundation/card-mode?category=ci-pipeline-orchestration

**Comment by alxfu** [comment]
Jenkins https://github.com/dagger/dagger/discussions/1677

**Reply by tedandrew** [reply]
Especially Jenkins on Kubernetes!

**Comment by thatsk** [comment]
Jenkins

**Comment by Cwagne17** [comment]
Circle ci!

**Comment by vdonkor** [comment]
Jenkins

**Comment by fearless23** [comment]
AWS CodePipelines

**Comment by Scizoo88** [comment]
drone.io

**Comment by peijianju-sonrai** [comment]
flux v1 and v2

**Comment by askkhan84** [comment]
AWS codepipeline

**Comment by marsmike** [comment]
Jenkins please. 👍

**Comment by shivakalva** [comment]
Jenkins

**Comment by a-stein** [comment]
Bamboo, please

**Reply by satyamkapoor** [reply]
Would be nice to have it but honestly companies using Bamboo are so dated in tech to decide quickly to move to dagger :(

**Comment by crispibits-fr** [comment]
Jenkins please

**Comment by kyapwc** [comment]
CircleCI please

**Comment by ciaransweet** [comment]
As much as I'd love to be back on GitHub Actions, **Jenkins** please!

**Reply by antikaspar** [reply]
Jenkins please

**Comment by PPHyperion** [comment]
[Drone CI](https://www.drone.io]) please

**Comment by chales** [comment]
Another vote for Jenkins.

**Comment by munichbughunter** [comment]
Jenkins please

**Comment by jbapu** [comment]
Jenkins +1

**Comment by chrisgrove-keysight** [comment]
Jenkins

**Comment by ihangji** [comment]
Jenkins

**Reply by OneCricketeer** [reply]
Five comments in a row say jenkins. Can we please search and simply upvote previous comments?

**Comment by mattiaforc** [comment]
Been using [GoCD](https://www.gocd.org/) with Buildah rootless builds for a while, GoCD would be a cool addition!

**Comment by drpdishant** [comment]
Tekton

**Comment by PashCracken** [comment]
TeamCity

**Reply by darkn3rd** [reply]
TeamCity support is important to my organization.

**Comment by michaelhthomas** [comment]
Azure DevOps

**Comment by pbadenski** [comment]
Semaphore

**Comment by TrymKristian** [comment]
Azure DevOps Server

**Comment by ljy545** [comment]
I am sending you 1π! Pi is a new digital currency developed by Stanford PhDs, with over 33 million members worldwide. To claim your Pi, follow this link https://minepi.com/wuhh123 and use my username (wuhh123) as your invitation code.If 1Pi=100000$ today, how would you arrange your digital assets?？

And I will not hesitate to use 20% to buy a car and a house, and then use 30 to invest, and the remaining 50% will continue to hold #PiNetwork until it is 10 times more and then make other plans.

**Comment by junkpiano** [comment]
[Bitrise](https://bitrise.io)

**Reply by sierisimo** [reply]
Yes please, their CLI is cool but the integration would be really helpful

**Comment by dnalbach** [comment]
Bitbucket Pipelines - I have lots of clients using this

**Reply by jpadams** [reply]
https://github.com/dagger/dagger/discussions/1677#discussioncomment-3195089

**Comment by dnalbach** [comment]
AWS CodeBuild/CodePipeline

**Comment by guoyu07** [comment]
[Drone](https://www.drone.io/)

**Comment by mudassir-ngineous** [comment]
Circleci

**Reply by jpadams** [reply]
https://github.com/dagger/dagger/discussions/1677#discussioncomment-3208937

**Comment by borisovano** [comment]
Argo stack (workflows, events and pipelines)

**Comment by alberto-cerato** [comment]
Zuul CI
https://zuul-ci.org/

**Comment by talk2drys** [comment]
Bitbucket Pipeline

**Reply by jpadams** [reply]
https://github.com/dagger/dagger/discussions/1677#discussioncomment-3195089

**Comment by Pandry** [comment]
[woodpecker](https://woodpecker-ci.org/)

**Comment by scheung38** [comment]
Devtron CI/CD

**Comment by darkn3rd** [comment]
I am interested in [Spinnaker](https://spinnaker.io/) CD pipelines, which are reflected in JSON, and can be downloaded.  They can also be uploaded with changes to create new pipelines.  There's a CLI tool that I use to manage the pipelines https://spinnaker.io/docs/guides/spin/pipeline/

**Reply by jpadams** [reply]
> I am interested in [Spinnaker](https://spinnaker.io/) CD pipelines, which are reflected in JSON, and can be downloaded. They can also be uploaded with changes to create new pipelines. There's a CLI tool that I use to manage the pipelines https://spinnaker.io/docs/guides/spin/pipeline/

Sounds like an ideal use for Dagger. You could consume/generate the Spinnaker JSON and invoke the CLI. Let me know if you'd like to set up a POC/pilot to get that going  :)

**Comment by dezren39** [comment]
nomad jobs please :-)

**Comment by Harrolee** [comment]
GCP Cloud Build 😁

**Reply by pfcarrier** [reply]
I second that, having one more example with GCP Cloud Build and AWS Code Build would complete the collection of all usual CI plus all the hyperscaller's. one.  This would create a nice "Wow" factor for any MSP that are revewing dagger since it would demonstrate all the CI of all their clients can leverage dagger and further demonstrate the joy dagger brings to the table.

**Comment by cgetzen** [comment]
I'd love to see CodeFresh support! I'm also curious if there's any documentation around how to contribute CI adapters ourselves.

**Reply by edromero** [reply]
It's not hard getting it working in Codefresh but it requires granting root privileges to the job. Currently only the hybrid option where you run the Codefresh workload on your own k8s cluster will allow this.
